src/main/scala/is/hail/utils/IntIterator.scala

```diff
@@ -0,0 +1,14 @@
+package is.hail.utils
+
+abstract class IntIterator extends Iterator[Int] {
+  def nextInt(): Int
+  override def next(): Int = nextInt()
+
+  override def count(p: Int => Boolean): Int = {
```

Comment:
`Int => Boolean` is a `Function1[Int, Boolean]`.  That means you allocate a Function1 object (per-variant, OK), but the both the Int and Boolean are boxed completely obviating the benefit of `IntIterator`.

In this case that isn't true since Function1 is specialized for in the Int => Boolean case, but not for all combinations of primitive types.  The specialization isn't even documented:

http://www.scala-lang.org/api/current/scala/Function1.html

you need to look at the code:

https://github.com/scala/legacy-svn-scala/blob/master/src/library/scala/Function1.scala#L29

so I think this sets a bad precedent.  My instinct is now "if it looks like a Scala collection it performs badly" and should not be used in per-genotype operations.